aten/src/ATen/native/mps/OperationUtils.mm

```diff
@@ -0,0 +1,365 @@
+}
+
+void MPSGeneratorImpl::set_state(const c10::TensorImpl& new_state) {
+  static const size_t seed_size = sizeof(uint64_t);
+  static const size_t offset_size = sizeof(int64_t);
+  static const size_t total_size = seed_size + offset_size;
+
+  detail::check_rng_state(new_state);
+
+  auto new_state_size = new_state.numel();
+
+  uint64_t input_seed;
+  auto new_rng_state = new_state.data<uint8_t>();
+  memcpy(&input_seed, new_rng_state, seed_size);
+  this->set_current_seed(input_seed);
+}
+
+MPSGeneratorImpl* MPSGeneratorImpl::clone_impl() const {
+  auto gen = new MPSGeneratorImpl(0);
+  gen->set_current_seed(this->seed_);
+  return gen;
+}
+
+void runMPSGraph(
+    MPSStream* mpsStream,
+    MPSGraph* mpsGraph,
+    NSDictionary* feeds,
+    NSDictionary* results) {
+
+  dispatch_sync(mpsStream->queue(), ^() {
+    @autoreleasepool {
+      mpsStream->commit(true);
+      id<MTLCommandQueue> commandQueue = mpsStream->commandQueue();
+      MPSGraphExecutionDescriptor *executionDescriptor = [[MPSGraphExecutionDescriptor new] autorelease];
+
+      executionDescriptor.completionHandler = ^(NSDictionary<MPSGraphTensor *,
+                                                MPSGraphTensorData *> * resultsDictionary,
+                                                NSError * _Nullable error) {
+      };
+
+      [mpsGraph runAsyncWithMTLCommandQueue:commandQueue
+                                feeds:feeds
+                     targetOperations:nil
+                    resultsDictionary:results
+                  executionDescriptor:executionDescriptor];
+
+    }
+  });
+}
+
+MPSDataType getMPSDataType(ScalarType scalar_type) {
+  switch (scalar_type) {
+    case ScalarType::Float:
+      return MPSDataTypeFloat32;
+    case ScalarType::Half:
+      return MPSDataTypeFloat16;
+    case ScalarType::Int:
+      return MPSDataTypeInt32;
+    case ScalarType::Long:
+      return MPSDataTypeInt64;
+    case ScalarType::Short:
+      return MPSDataTypeInt16;
+    case ScalarType::Byte:
+      return MPSDataTypeInt8;
+    case ScalarType::Bool:
+      return MPSDataTypeBool;
+    default:
+      TORCH_CHECK_TYPE(scalar_type == ScalarType::Double, "Unsupported data type ", scalar_type, " on MPS backend");
+      return MPSDataTypeFloat32;
+  }
+}
+
+MPSDataType getMPSScalarType(ScalarType scalar_type) {
+  switch (scalar_type) {
+    case ScalarType::Double:
+    case ScalarType::Float:
+      return MPSDataTypeFloat32;
+    case ScalarType::Half:
+      return MPSDataTypeFloat16;
+    case ScalarType::Int:
+      return MPSDataTypeInt32;
+    case ScalarType::Long:
+      return MPSDataTypeInt64;
+    case ScalarType::Short:
+      return MPSDataTypeInt16;
+    case ScalarType::Byte:
+      return MPSDataTypeInt8;
+    case ScalarType::Bool:
+      return MPSDataTypeBool;
+    default:
+      TORCH_CHECK_TYPE(false, "Unsupported data type '", scalar_type, "' on MPS backend");
+  }
+}
+
+std::string getMPSTypeString(ScalarType scalar_type) {
```

Comment:
it doesn't seem like this needs to return a std::string, const char* will work perfectly well